tb/eltwise_top.cpp

```diff
@@ -1,27 +1,34 @@
-#include <hls_stream.h>
-using namespace hls;
-#include "ap_int.h"
 #include "eltwise.hpp"
 #include "interpret.hpp"
 
+#include <cassert>
 
 #include "data/eltwise_config.h"
 
-void Testbench_Eltwise(int mode, stream<ap_uint<NUM_CHANNELS * INPUT_1_WIDTH> > & in0, stream<ap_uint<NUM_CHANNELS * INPUT_2_WIDTH> > & in1, stream<ap_uint<NUM_CHANNELS * OUTPUT_WIDTH> > & out) 
-{
-    switch(mode) {
-		case 0:
-			AddEltwiseFunction<ap_int<INPUT_1_WIDTH>, ap_int<INPUT_2_WIDTH>, ap_int<OUTPUT_WIDTH>> add_eltwise_fxn;
-            StreamingEltwise<NUM_CHANNELS, NUM_CHANNELS, NUM_WORDS, Slice<ap_int<INPUT_1_WIDTH>>, Slice<ap_int<INPUT_2_WIDTH>>, Slice<ap_int<OUTPUT_WIDTH>> >(in0, in1, out, add_eltwise_fxn);
-			break;
-		case 1:
-			SubEltwiseFunction<ap_int<INPUT_1_WIDTH>, ap_int<INPUT_2_WIDTH>, ap_int<OUTPUT_WIDTH>> sub_eltwise_fxn;
-            StreamingEltwise<NUM_CHANNELS, NUM_CHANNELS, NUM_WORDS, Slice<ap_int<INPUT_1_WIDTH>>, Slice<ap_int<INPUT_2_WIDTH>>, Slice<ap_int<OUTPUT_WIDTH>> >(in0, in1, out, sub_eltwise_fxn);
-			break;
-		case 2:
-			AbsDiffEltwiseFunction<ap_int<INPUT_1_WIDTH>, ap_int<INPUT_2_WIDTH>, ap_int<OUTPUT_WIDTH>> absdiff_eltwise_fxn;
-            StreamingEltwise<NUM_CHANNELS, NUM_CHANNELS, NUM_WORDS, Slice<ap_int<INPUT_1_WIDTH>>, Slice<ap_int<INPUT_2_WIDTH>>, Slice<ap_int<OUTPUT_WIDTH>> >(in0, in1, out, absdiff_eltwise_fxn);
-		default:
-            break;	
+
+void Testbench_Eltwise(
+	int  mode,
+	hls::stream<ap_uint<NUM_CHANNELS * INPUT_1_WIDTH>> &in0,
+	hls::stream<ap_uint<NUM_CHANNELS * INPUT_2_WIDTH>> &in1,
+	hls::stream<ap_uint<NUM_CHANNELS * OUTPUT_WIDTH>>  &out
+) {
+	switch(mode) {
+	case 0:
+		StreamingEltwise<NUM_CHANNELS, NUM_CHANNELS, NUM_WORDS, Slice<ap_int<INPUT_1_WIDTH>>, Slice<ap_int<INPUT_2_WIDTH>>, Slice<ap_int<OUTPUT_WIDTH>>>(
+			in0, in1, out, [](auto a, auto b) { return  a + b; }
+		);
+		break;
+	case 1:
+		StreamingEltwise<NUM_CHANNELS, NUM_CHANNELS, NUM_WORDS, Slice<ap_int<INPUT_1_WIDTH>>, Slice<ap_int<INPUT_2_WIDTH>>, Slice<ap_int<OUTPUT_WIDTH>>>(
+			in0, in1, out, [](auto a, auto b) { return  a - b; }
+		);
+		break;
+	case 2:
+		StreamingEltwise<NUM_CHANNELS, NUM_CHANNELS, NUM_WORDS, Slice<ap_int<INPUT_1_WIDTH>>, Slice<ap_int<INPUT_2_WIDTH>>, Slice<ap_int<OUTPUT_WIDTH>>>(
+			in0, in1, out, [](auto a, auto b) { return  a>b? a-b : b-a; }
+		);
+		break;
```

Comment:
is it possible to make the lambda function a variable to avoid code duplication for the StreamingEltwise function call, so that it can be factored out, something similar to

```
auto lambda_fxn;
lambda_fxn = [](auto a, auto b) { return  a>b? a-b : b-a; }
StreamingEltwise<NUM_CHANNELS, NUM_CHANNELS, NUM_WORDS, Slice<ap_int<INPUT_1_WIDTH>>, Slice<ap_int<INPUT_2_WIDTH>>, Slice<ap_int<OUTPUT_WIDTH>>>(
			in0, in1, out, lambda_fxn
		);
```

Reply:
I don't really see how this makes it simpler but sure, this is is possible. You only need to make an instant assignment to the `auto`-typed variable so that the compiler can actually infer its type:
```C++
auto const  lambda_fxn = [](auto a, auto b) { return  a>b? a-b : b-a; };
```

Reply:
Sorry I see that my code example there wasn't sufficient as an example, what I meant was to do something like this:

```
auto lambda_fxn:
switch(mode) {
case 0: lambda_fxn = [](auto a, auto b) { return  a+b; }; break;
case 1:  lambda_fxn = [](auto a, auto b) { return  a-b; }; break;
// and so on
}
StreamingEltwise<NUM_CHANNELS, NUM_CHANNELS, NUM_WORDS, Slice<ap_int<INPUT_1_WIDTH>>, Slice<ap_int<INPUT_2_WIDTH>>, Slice<ap_int<OUTPUT_WIDTH>>>(
			in0, in1, out, lambda_fxn
		);
```

Reply:
No, you don't want to do this. This would require a `lambda_fxn` variable of some common supertype with a virtual dispatch to the implemented operator. Such a construct would no longer synthesize.
You really want the literal lambda expression to instantiate its very own `StreamingEltwise<>` template specialization where its behavior is inlined.

Reply:
Okay, then let's keep it like this. No further comments from me.